vlog4j-client/src/main/java/org/vlog4j/client/picocli/VLog4jClientMaterialize.java

```diff
@@ -0,0 +1,187 @@
+
+import org.semanticweb.vlog4j.core.model.api.PositiveLiteral;
+import org.semanticweb.vlog4j.core.reasoner.Algorithm;
+import org.semanticweb.vlog4j.core.reasoner.KnowledgeBase;
+import org.semanticweb.vlog4j.core.reasoner.LogLevel;
+import org.semanticweb.vlog4j.core.reasoner.Reasoner;
+import org.semanticweb.vlog4j.core.reasoner.implementation.VLogReasoner;
+import org.semanticweb.vlog4j.examples.ExamplesUtils;
+import org.semanticweb.vlog4j.parser.ParsingException;
+import org.semanticweb.vlog4j.parser.RuleParser;
+
+import picocli.CommandLine.ArgGroup;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+
+@Command(name = "materialize", description = "Execute the chase and store the literal's extensions")
+public class VLog4jClientMaterialize implements Runnable {
+
+	// TODO add link to rls syntax
+	@Option(names = "--rule-file", description = "Rule file(s) in rls syntax", required = true)
+	private List<String> ruleFiles = new ArrayList<>();
+
+//  TODO
+//	Support graal rule files
+//	@Option(names = "--graal-rule-file", description = "Rule file(s) in graal syntax", required = true)
+//	private List<String> graalRuleFiles = new ArrayList<>();
+
+	@Option(names = "--log-level", description = "Log level of VLog (c++ library). One of: DEBUG, INFO, WARNING (default), ERROR.", required = false)
+	private LogLevel logLevel = LogLevel.WARNING;
+
+	@Option(names = "--log-file", description = "Log file of VLog (c++ library). VLog will log to the default system output by default", required = false)
+	private String logFile;
+
+	@Option(names = "--chase-algorithm", description = "Chase algorithm. RESTRICTED_CHASE (default) or SKOLEM_CHASE.", required = false)
+	private Algorithm chaseAlgorithm = Algorithm.RESTRICTED_CHASE;
+
+	@Option(names = "--timeout", description = "Timeout in seconds. Infinite by default", required = false)
+	private int timeout = 0;
+
+	@Option(names = "--query", description = "Positive not-ground Literals to query after materialization in rls syntax. Vlog4jClient will print the size of its extension", required = true)
+	private List<String> queryStrings = new ArrayList<>();
+
+	@ArgGroup(exclusive = false)
+	private PrintQueryResults printQueryResults = new PrintQueryResults();
+
+	@ArgGroup(exclusive = false)
+	private SaveQueryResults saveQueryResults = new SaveQueryResults();
+
+	// TODO save model
+	// @ArgGroup(exclusive = false)
+	// private SaveModel saveModel = new SaveModel();
+
+	@Override
+	public void run() {
+		ExamplesUtils.configureLogging();
+
+		if (!printQueryResults.isConfigOk()) {
+			printQueryResults.printErrorAndExit();
+		}
+
+		if (!saveQueryResults.isConfigOk()) {
+			saveQueryResults.printErrorAndExit();
+		}
+
+		// TODO
+		// if (!saveModel.isConfigOk()) {
+		// saveModel.printErrorAndExit();
+		// }
+
+		System.out.println("Configuration:");
+
+		/* Configure rules */
+		final KnowledgeBase kb = new KnowledgeBase();
+		for (String ruleFile : ruleFiles) {
+			try {
```

Comment:
the best practice is to have only the statements that generate exceptions surrounded by try-catch. Also, you can separate the two statements that throw two different exceptions here on two lines.

```
final FileInputStream fileInputStream;
try{fileInputStream = FileInputStream(ruleFile);
} catch(FileNotFoundException  e){
...
}
try{RuleParser.parseInto(kb, fileInputStream);
} catch ....
```

Reply:
Maybe I am missing something here, but I don't think this would work because the variable `fileInputStream` won't be defined in scope of the second try.

Reply:
You are right, you would have to initialise the variable, because the compiler does not know that if the first exception is thrown, then the program will terminate due to System.exit.
Then the code would look like below, which is not very nice, either. You can keep your version, but move the System.out out of the try{} block. 

```
for (final String ruleFile : this.ruleFiles) {
			FileInputStream fis = null;
			try {
				fis = new FileInputStream(ruleFile);

			} catch (final FileNotFoundException e) {
				System.err.println("File not found: " + ruleFile + ". " + e.getMessage());
				System.exit(1);
			}
			try {
				RuleParser.parseInto(kb, fis);
			} catch (final ParsingException e) {
				System.err.println("Failed to parse rule file: " + ruleFile + ". " + e.getMessage());
				System.exit(1);
			}
			System.out.println("  --rule-file: " + ruleFile);
		}
```